### PR TITLE
[go_router] Feat add route redirect shellroutes (#114559)

### PR DESCRIPTION
**Pull request by VictorOhashi**
Possible solution for: https://github.com/flutter/flutter/issues/114559

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I [linked to at least one issue that this PR fixes] in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

### CODE REVIEW

**Comment by cedvdb** [comment]
@VictorOhashi I don't think the reviewer works anymore, you might want to tag someone else.

**Comment by VictorOhashi** [comment]
@johnpryan could you have a look?

File: packages/go_router/lib/src/configuration.dart [hunk 1]
```diff
@@ -434,18 +433,18 @@ class RouteConfiguration {
...
-    final RouteMatch match = routeMatches[currentCheckIndex];
+    final RouteMatchBase match = routeMatches[currentCheckIndex];
     FutureOr<String?> processRouteRedirect(String? newLocation) =>
         newLocation ??
         _getRouteLevelRedirect(
             context, matchList, routeMatches, currentCheckIndex + 1);
-    final GoRoute route = match.route;
+    final RouteBase route = match.route;
     FutureOr<String?> routeRedirectResult;
     if (route.redirect != null) {
```
Comment (chunhtai) [inline packages/go_router/lib/src/configuration.dart:449]: This must not be null now?

**Review by chunhtai — COMMENTED** [review]
Thanks for your contribution, but we need a test to verify the new feature. Can you also bump the minor version since this is a new api?

**Comment by VictorOhashi** [comment]
> Thanks for your contribution, but we need a test to verify the new feature. Can you also bump the minor version since this is a new api?

done

File: packages/go_router/lib/src/configuration.dart (hunk 1 above)
Comment (chunhtai) [reply packages/go_router/lib/src/configuration.dart:449]: yes, but I meant you can remove this conditional check.

File: packages/go_router/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+## 13.3.0
+
+- Feat add route redirect to ShellRoutes
```
Comment (chunhtai) [inline packages/go_router/CHANGELOG.md:3]:
```suggestion
- Adds route redirect to ShellRoutes
```

**Review by chunhtai — COMMENTED** [review]
It is still missing a test

File: packages/go_router/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## 14.0.1
```
Comment (johnpryan) [inline packages/go_router/CHANGELOG.md:1]: `14.1.0`, since this adds a new API?

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
looks like there is a conflict

**Review by hangyujin — APPROVED** [review]
LGTM

**Pull request merged**